modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SSource.java

```diff
@@ -42,63 +45,51 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 
 import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableMap;
 import static org.elasticsearch.painless.WriterConstants.BASE_CLASS_TYPE;
 import static org.elasticsearch.painless.WriterConstants.CLASS_TYPE;
 import static org.elasticsearch.painless.WriterConstants.CONSTRUCTOR;
-import static org.elasticsearch.painless.WriterConstants.EXECUTE;
-import static org.elasticsearch.painless.WriterConstants.MAP_GET;
-import static org.elasticsearch.painless.WriterConstants.MAP_TYPE;
 
 /**
  * The root of all Painless trees.  Contains a series of statements.
  */
 public final class SSource extends AStatement {
 
     /**
-     * Tracks reserved variables.  Must be given to any source of input
+     * Tracks derived arguments and the loop counter.  Must be given to any source of input
      * prior to beginning the analysis phase so that reserved variables
      * are known ahead of time to assign appropriate slots without
      * being wasteful.
      */
     public interface Reserved {
-        void markReserved(String name);
-        boolean isReserved(String name);
+        void markUsedVariable(String name);
 
         void setMaxLoopCounter(int max);
         int getMaxLoopCounter();
     }
 
     public static final class MainMethodReserved implements Reserved {
-        private boolean score = false;
-        private boolean ctx = false;
+        private final Map<String, DerivedArgument> usedDerivedArguments = new LinkedHashMap<>();
```

Comment:
Note that we might change the order of the derived argument now - it used to be they'd be in the declared order. Now they are in the order of first used (as seen by walking the AST).